src/forms/create_ticket.ts

```diff
@@ -20,3 +25,189 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
+    async getCreateTicketFields(): Promise<AppField[]> {
+        const zdTicketForms = await tryPromiseWithMessage(this.zdClient.ticketforms.list(), 'Failed to fetch ticket forms');
+
+        // show the form selector when intially opening the modal
+        this.addFormsSelectField(zdTicketForms);
+
+        // only show form select field until user selects a form select value
+        if (!this.currentFieldValuesSelected()) {
+            return this.appFields;
+        }
+
+        // get the form id from the selected form field value
+        const formIDSelectValue = this.mapper.getSavedValue(AppFieldNames.FormsSelectName);
+        const formID = formIDSelectValue.value;
+
+        const zdFormFieldIDs = this.getTicketFieldIDs(zdTicketForms, formID);
+        const zdTicketFields = await tryPromiseWithMessage(this.zdClient.ticketfields.list(), 'Failed to fetch ticket fields');
+        const zdViewableFields = this.getViewableFields(zdTicketFields, zdFormFieldIDs);
+
+        this.mapZdFieldsToAppFields(zdViewableFields);
+
+        return this.appFields;
+    }
+
+    // getTicketFieldIDs returns the list of all fields in a Zendesk form
+    private getTicketFieldIDs(forms: any, id: number): number[] {
+        const ids = forms.find((form: any) => {
+            return form.id.toString() === id;
+        }).ticket_field_ids;
+        return ids;
+    }
+
+    // getViewableFields returns a list of viewable Zendesk field IDs
+    private getViewableFields(ticketFields: UserField[], formIDs: number[]): Tickets.Field[] {
+        const fields = [];
+        ticketFields.forEach((field: UserField) => {
+            // omit fields that do not show up in the create ticket modeal in Zendesk
+            // but are returned in the ticketFields query
+            const omitFields = ['Group', 'Status'];
+            if (omitFields.includes(field.title)) {
+                return;
+            }
+
+            // only keep fields listed in formIDs
+            if (formIDs.includes(field.id)) {
+                fields.push(field);
+            }
+        });
+        return fields;
+    }
+
+    mapZdFieldsToAppFields(fields: UserField[]): AppField[] {
+        const appFields: AppField[] = [];
+        fields.forEach((field) => {
+            const appField: AppField = {
+                name: this.getMappedName(field),
+                label: field.title,
+                subtype: '',
+                is_required: Boolean(field.required_in_portal),
+            };
+
+            switch (field.type) {
+            case ZDFieldTypes.Description:
+                // will be filled by post message and handled separately
+                return;
+
+            case ZDFieldTypes.Integer:
+            case ZDFieldTypes.Decimal: {
+                this.addField(this.mapper.mapToText(appField));
+                return;
+            }
+            case ZDFieldTypes.Subject:
+            case ZDFieldTypes.Text:
+            case ZDFieldTypes.MultiLine: {
```

Comment:
Yes, this would work if the mapper provided newTextField(), newSelectField() methods with submethods to modify the resulting field type.

```javascript
const field = newTextField(<required-field-inputs>)
field.setMaxLength(2048)
field.setMaxLength(1)
field.isRequired(true)
this.mapper.addField(field)
```

This would actually be better for development because the inputs required and optional for each field won't need to be guessed.  They are hardcoded as inputs to the class constructor

Reply:
To acheive, remove addBoolField, addTextField, .etc out of the mapper (rename mapper to form_builder)
create new class fields with newBoolField, newTextField, .etc helpers. 
- required each newXxxField() to have required inputs that are needed. Thes can be the base class that includes required, shared fields among all field types